File: content/thoughts/university/twenty-three-twenty-four/sfwr-3bb4/a3/Quicksort.java
import java.util.Random;

public class Quicksort {
    static int N;   // number of elements to be sorted
    static int S;   // threshold for creating a sub-thread
    static int a[]; // array to be sorted
    static int activeThreads = 0;

    static int partition(int p, int r) {
        int x = a[r];
        int i = p - 1;
        for (int j = p; j <= r - 1; j++) {
            if (a[j] <= x) {
                i++; 
                int t = a[i]; a[i] = a[j]; a[j] = t;
            }
        }
        int t = a[i + 1]; a[i + 1] = a[r]; a[r] = t;
        return i + 1;
    }

    static void sequentialsort(int p, int r) {
        if (p < r) {
            int q = partition(p, r);
            sequentialsort(p, q - 1);
            sequentialsort(q + 1, r);
        }
    }

    static void parallelsort(int p, int r) {
        if (p < r) {
            int q = partition(p, r);
            int left = q - p;
            int right = r - q;
    
            if (left > S && right > S && activeThreads < S) {
                Thread rightThread = new Thread(() -> {
                    activeThreads++;
                    parallelsort(q + 1, r);
                    activeThreads--;
                });
    
                rightThread.start();
                parallelsort(p, q - 1);
    
                try {
                    rightThread.join();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            } else {
                sequentialsort(p, q - 1);
                sequentialsort(q + 1, r);
            }
        }
    }

    public static void main(String args[]) {
        N = Integer.parseInt(args[0]);
        S = Integer.parseInt(args[1]);
        a = new int[N];
        Random random = new Random();
        for (int i = 0; i < N; i++) a[i] = random.nextInt(10000);

        final long start = System.currentTimeMillis();
        parallelsort(0, N - 1);
        final long end = System.currentTimeMillis();
        
        for (int i = 1; i < N; i++) assert a[i - 1] <= a[i];
        System.out.println((end - start) + " ms");
    }
}

File: content/dump/university/sfwr-3bb4/a3/quicksort.py
import os
from threading import Thread, Semaphore
from random import randint
from time import time

MAX_THREADS = int(os.environ.get('MAX_THREADS', 100))
sem = Semaphore(MAX_THREADS)

def partition(p, r): 
  x, i = a[r], p - 1
  for j in range(p, r):
    if a[j] <= x: 
      i += 1; a[i], a[j] = a[j], a[i]
  a[i + 1], a[r] = a[r], a[i + 1]
  return i + 1;
    

def sequentialsort(p, r):
  if p < r:
    q = partition(p, r)
    sequentialsort(p, q - 1)
    sequentialsort(q + 1, r)

def parallelsort(p, r):
  if p<r:
    q = partition(p, r)
    left = q - p
    right = r - q
    if left > S and right > S:
      rthread = Thread(target=parallelsort, args=(q + 1, r))
      if sem.acquire(blocking=False):
        rthread.start()
        parallelsort(p, q - 1)
        rthread.join()
        sem.release()
      else:
        sequentialsort(p, r)
    else:
      sequentialsort(p, q-1)
      sequentialsort(q+1, r)

def quicksort(N, step):
  global a, S
  a, S = [randint(0, 10000) for _ in range(N)], step

  start = time()    
  parallelsort(0, N - 1)
  end = time()

  print(str(int((end - start) * 1000)) + " ms")
  for i in range(1, N): assert a[i - 1] <= a[i]    
